www-apps/bugzilla/bugzilla-5.0.6.ebuild

```diff
@@ -0,0 +1,83 @@
+# Copyright 1999-2020 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
+
+inherit webapp
+
+DESCRIPTION="Bugzilla is the Bug-Tracking System from the Mozilla project"
+SRC_URI="https://ftp.mozilla.org/pub/mozilla.org/webtools/${P}.tar.gz"
+HOMEPAGE="https://www.bugzilla.org"
+
+LICENSE="MPL-2.0"
+KEYWORDS="~amd64 ~x86"
+
+IUSE="mysql postgres sqlite test"
+RESTRICT="!test? ( test )"
+REQUIRED_USE="|| ( mysql postgres sqlite )"
+
+# sorting is identical to upstream MYMETA.json, please don't change
+# Theses are mandatory for checksetup.pl to configure bugzilla
+BASIC_DEPS="
```

Comment:
`Theses are mandatory for checksetup.pl`
So why name it `BASIC_DEPS` instead of calling it `DEPEND`?

There's a typo btw.

Reply:
Right! I named it BASIC_DEPS while waiting for what to do about optional runtime dependencies. In fact BASIC_DEPS only need to be in RDEPEND and not in DEPEND (I think DEPEND is not often used for web applications).

@thesamesam how about I write an ebuild for a www-apps/bugzilla-extras which would DEPEND on www-apps/bugzilla and in which I introduce all optional dependencies and the jobqueue openrc startup file? (So that we always have a simple yet fully functionnal bugzilla package) In this way I follow what I understand is done with LaTeX packages and it would also close https://bugs.gentoo.org/332251.

Reply:
There are about 35 packages IIRC, which seems to be misc utilities (I have two fully functional installs at home that have been running for at least a month). I'll wait for the best solution :-)

Reply:
Can you list the deps and how they work? If they are pure runtime optional dependencies, you should use `optfeature` from *eutils.eclass* to control them.

You can perhaps condense some dependencies under common USE flag, but let's hear the list if that's possible.

Ie, what dep is required for what functionality, and is it pure optional dependency or needed by something else? And does the `checksetup.pl` only work when configuring bugzilla the first time?

Sorry I've never installed bugzilla before.

Reply:
When you install bugzilla, you need to run checksetup.pl. This program lists you 37 optional perl-modules with the feature they enable. I thought that 37 is a big number of packages to list as optfeature and is a bit redundant. 

But I can re-add these deps with an extra use flag if it is policy-compliant :-) Just waiting for the appropriate decision (extras use flag for optional runtime deps or another package or something else ?)

Reply:
The following Perl modules are optional:
Checking for                   GD (v1.20)     
Checking for                Chart (v2.4.1)    
Checking for          Template-GD (any)       
Checking for           GDTextUtil (any)       
Checking for              GDGraph (any)       
Checking for           MIME-tools (v5.406)    
Checking for          libwww-perl (any)       
Checking for             XML-Twig (any)       
Checking for          PatchReader (v0.9.6)    
Checking for            perl-ldap (any)       
Checking for          Authen-SASL (any)       
Checking for         Net-SMTP-SSL (v1.01)     
Checking for           RadiusPerl (any)       
Checking for            SOAP-Lite (v0.712)    
Checking for          XMLRPC-Lite (v0.712)    
Checking for             JSON-RPC (any)       
Checking for           Test-Taint (v1.06)     
Checking for          HTML-Parser (v3.67)     
Checking for        HTML-Scrubber (any)       
Checking for               Encode (v2.21)     
Checking for        Encode-Detect (any)       
Checking for          Email-Reply (any)       
Checking for HTML-FormatText-WithLinks (v0.13)
Checking for          TheSchwartz (v1.07)     
Checking for       Daemon-Generic (any)       
Checking for             mod_perl (v1.999022) 
Checking for     Apache-SizeLimit (v0.96)     
Checking for        File-MimeInfo (any)       
Checking for           IO-stringy (any)       
Checking for      Cache-Memcached (any)       
Checking for  File-Copy-Recursive (any)       
Checking for           File-Which (any)       
Checking for              mod_env (any)       
Checking for          mod_expires (any)       
Checking for          mod_headers (any)       
Checking for          mod_rewrite (any)       
Checking for          mod_version (any)

Reply:
I forgot: the last 5 are apache modules. The 4 first modules are installed by default, and mod_version is not needed as we do not have Apache2.2 so we do not need to worry about those. There are 3 optional perl modules that are installed by default because of rdeps. That leaves 29 optional pure perl modules for which checksetup.pl says:

*               MODULE NAME * ENABLES FEATURE(S)                      *                                                        
***********************************************************************
*                        GD * Graphical Reports, New Charts, Old Charts *
*                     Chart * New Charts, Old Charts                  *
*               Template-GD * Graphical Reports                       *
*                GDTextUtil * Graphical Reports                       *
*                   GDGraph * Graphical Reports                       *
*                MIME-tools * Move Bugs Between Installations         *
*               libwww-perl * Automatic Update Notifications          *
*                  XML-Twig * Move Bugs Between Installations, Automatic Update Notifications *
*               PatchReader * Patch Viewer                            *
*                 perl-ldap * LDAP Authentication                     *
*               Authen-SASL * SMTP Authentication                     *
*              Net-SMTP-SSL * SSL Support for SMTP                    *
*                RadiusPerl * RADIUS Authentication                   *
*                 SOAP-Lite * XML-RPC Interface                       *
*               XMLRPC-Lite * XML-RPC Interface                       *
*                  JSON-RPC * JSON-RPC Interface, REST Interface      *
*                Test-Taint * JSON-RPC Interface, XML-RPC Interface, REST Interface *
*             HTML-Scrubber * More HTML in Product/Group Descriptions *
*             Encode-Detect * Automatic charset detection for text attachments *
*               Email-Reply * Inbound Email                           *
* HTML-FormatText-WithLinks * Inbound Email                           *
*               TheSchwartz * Mail Queueing                           *
*            Daemon-Generic * Mail Queueing                           *
*                  mod_perl * mod_perl                                *
*          Apache-SizeLimit * mod_perl                                *
*             File-MimeInfo * Sniff MIME type of attachments          *
*                IO-stringy * Sniff MIME type of attachments          *
*           Cache-Memcached * Memcached Support                       *
*       File-Copy-Recursive * Documentation                           *
***********************************************************************

Also: Net-SMTP-SSL is deprecated (https://metacpan.org/pod/Net::SMTP::SSL, https://bugzilla.mozilla.org/show_bug.cgi?id=1316255)

Reply:
So if I count right, that'd be ~19 `optfeature` lines. Perhaps few less. I'd personally prefer that option instead of introducing a single USE flag to enable them **all**. 

What did you mean by a new package? How'd that work? What kind of installed file list would `bugzilla-extras` have?

Does any of these modules require any kind of customization on `bugzilla` package's list of installed files?

Reply:
29 not 19 ? (minus Net-SMTP-SSL that makes 28). One can also install media-gfx/graphviz as an optional runtime dependency, which I forgot... I quite like the optfeature option, as I am not at all familiar with all those optional deps. A package named bugzilla-extras could BDEPEND on bugzilla and RDEPEND on all those extra packages for instance.

Of all the optional deps, I am only familiar with TheSchwartz, which enables queuing mail via a perl script named jobqueue.pl (https://bugzilla.readthedocs.io/en/5.0/administering/parameters.html#email). This can be turned into a service via an openrc service file that can be installed by the ebuild as it was done in the 5.0.3-r2 ebuild. This is where it gets tricky: this openrc service file needs to know where bugzilla is installed, which may not be obvious. So the user has to create a .confd file specifying where the bugzilla install is located. This could easily be explained in an elog message.

For all the optional post-install configuration, the postinstall could also refer to https://bugzilla.readthedocs.io/en/5.0/installing/optional-post-install-config.html, which is quite good.

Reply:
Ok, reconfiguring files wrt the install dir may not be as tricky as I thought. Let me think about that.

Reply:
Problem is: I cannot rewrite a file that has been installed in /etc/conf.d (sandbox violation). I would need webapp-config to do that (it knows where bugzilla is installed). I guess I'm stuck with an elog message explaining to the user that one needs to create a one line conf.d file with the install path of bugzilla.

Reply:
I think you should be able to modify /etc/conf.d just fine. Just don't do /etc/conf.d *directly*? Users will need to merge your changes and theirs using e.g. dispatch-conf.

Reply:
I can't do it in the ebuild: if the package is emerged with vhosts flag (probably the recommended way), the app is installed with webapp-config at the users' preferred location. It is at this point that a file recording the installation path must be written in /etc/conf.d/. But then, thinking about it, multiple installations also require multiple startup files so webapp-config should also add a file in /etc/init.d/. Is that recommended?

I'm thinking that upon emerging the ebuild could unconditionnaly install the openrc init script, then invite the user to consider more features with ~29 optfeature lines. Then upon installing with webapp-config I could use the postinstall file to explain about the /etc/conf.d file, refer to the official bugzilla optional postinstall config, and to the gentoo wiki.

Reply:
I think what you're suggesting sounds good.